### ISSUE TITLE
java.lang.StackOverflowError for the entire cluster

### ISSUE THREAD

**Issue by hbrxa** [elastic/elasticsearch#25502]
Hmm... could be related to issue #24553 - looks somehow similar...

We are running a single cluster and Elastic crashed 5 times today. Is it possible to identify the query causing the bug?

**Elasticsearch version:**
5.4.2

**Plugins installed:**
analysis-icu
analysis-stempel
analysis-kuromoji
analysis-smartcn
readonlyrest.zip

**JVM version (java -version):**
1.8.0_131

**OS version (uname -a):**
Linux 3.19.0-37-generic 42~14.04.1-Ubuntu SMP Mon Nov 23 15:13:51 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

**Log Messages**
```
[2017-06-30T10:43:19,296][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [] fatal error in thread [elasticsearch[jRsKXqS][search][T#23]], exiting
java.lang.StackOverflowError: null
    at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1293) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]
    at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1293) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]
    at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1293) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]
    ...
```

**Comment by s1monw** [comment]
do you have the entire stacktrace?

**Comment by s1monw** [comment]
and are you using any regexp queries by any chance?

**Comment by hbrxa** [comment]
No... see my answer in #24553. We do not use any regexp queries.

**Comment by nik9000** [comment]
It'd be nice to get the rest of the stacktrace though.

**Comment by jasontedor** [comment]
The JVM limits the stack trace to 1024 elements by default. With a large thread stack, this is unlikely to cover the full stack trace and is mostly going to consist only of the recursive calls in the frame missing the bottom of the stack that we are looking for here. To get the bottom of the stack, the JVM needs to be started with a flag to not limit the depth of the stack trace to the default. This can be achieved with the JVM option: `-XX:MaxJavaStackTraceDepth=-1`; here, `-1` means unlimited which would guarantee us the full stack trace at the expense of excessive logout.

@hbrxa Would you be able to temporarily enable this via jvm.options until we can collect a full stack trace?

**Comment by hbrxa** [comment]
We are running elastic in production and avoid the error now: We tested everything locally but could not find the problematic queries. The solution was to rewrite the code by replacing all regular query_strings by simple_query_strings. So we assume the error was caused by some special characters in the search strings...

**Comment by hbrxa** [comment]
Today the error occured again - although using only simple_query_strings:

```
[2017-07-02T15:42:33,504][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [] fatal error in thread [elasticsearch[jRsKXqS][search][T#27]], exiting
java.lang.StackOverflowError: null
    at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1293) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]
    at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1293) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]
    at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1293) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]
    at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1293) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]
    at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1293) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]

```

**Comment by hbrxa** [comment]
Here is the bottom of the stack:

```
    at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1293) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]
    at org.apache.lucene.util.automaton.Operations.topoSortStates(Operations.java:1267) ~[lucene-core-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi - 2017-04-21 12:17:15]
    at org.apache.lucene.search.suggest.document.CompletionTokenStream.replaceSep(CompletionTokenStream.java:234) ~[lucene-suggest-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi -
2017-04-21 12:18:13]
    at org.apache.lucene.search.suggest.document.CompletionTokenStream.toAutomaton(CompletionTokenStream.java:182) ~[lucene-suggest-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi -
2017-04-21 12:18:13]
    at org.apache.lucene.search.suggest.document.CompletionTokenStream.toAutomaton(CompletionTokenStream.java:150) ~[lucene-suggest-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi -
2017-04-21 12:18:13]
    at org.apache.lucene.search.suggest.document.PrefixCompletionQuery.createWeight(PrefixCompletionQuery.java:71) ~[lucene-suggest-6.5.1.jar:6.5.1 cd1f23c63abe03ae650c75ec8ccb37762806cc75 - jimczi -
2017-04-21 12:18:13]
    at org.elasticsearch.search.suggest.completion.CompletionSuggester.suggest(CompletionSuggester.java:156) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.search.suggest.completion.CompletionSuggester.innerExecute(CompletionSuggester.java:70) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.search.suggest.completion.CompletionSuggester.innerExecute(CompletionSuggester.java:53) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.search.suggest.Suggester.execute(Suggester.java:39) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.search.suggest.SuggestPhase.execute(SuggestPhase.java:64) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:96) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:247) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:262) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:339) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.action.search.SearchTransportService$6.messageReceived(SearchTransportService.java:336) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.transport.TransportService$7.doRun(TransportService.java:627) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:638) ~[elasticsearch-5.4.2.jar:5.4.2]
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-5.4.2.jar:5.4.2]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_131]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_131]
    at java.lang.Thread.run(Thread.java:748) [?:1.8.0_131]

```

**Comment by hbrxa** [comment]
@jasontedor: We temporarily enable the jvm.option. Any idea what is causing this error... Would be great if the corresponding query could be logged...

Here an extract of our log-file when Elastic was restarted:

```
Starting program at:  Mon Jul  3 08:27:01 CEST 2017
Starting program at:  Mon Jul  3 10:21:01 CEST 2017
Starting program at:  Tue Jul  4 06:46:01 CEST 2017
Starting program at:  Tue Jul  4 15:17:01 CEST 2017
Starting program at:  Tue Jul  4 16:23:01 CEST 2017
Starting program at:  Tue Jul  4 20:39:01 CEST 2017
Starting program at:  Tue Jul  4 22:18:01 CEST 2017
Starting program at:  Wed Jul  5 06:39:01 CEST 2017
Starting program at:  Wed Jul  5 11:45:01 CEST 2017

```

**Comment by jasontedor** [comment]
@jimczi Could you take a look at this one?

**Comment by jimczi** [comment]
The stack trace indicates that the issue is related to the `suggest` query and not the`simple_query_string`.
The `suggest` query can also be affected by stack overflows but as the query_string only if you use very long suggest strings or explicitely build `regex` suggest query. 
@hbrxa could you paste here the gist of your suggest query ?

**Comment by jimczi** [comment]
We've put safeguards in Lucene to avoid StackOverflowError for large automaton and throw an IllegalArgumentException instead (which fails the query but does not kill the entire node).
The safeguards are going to be available in es 6.0 (Lucene 7):
https://issues.apache.org/jira/browse/LUCENE-7914 
The queries that failed should be very long queries so you can try to limit the size of the suggest queries to some reasonable size (a few hundreds chars should be enough) before upgrading to 6 when it's available. I'll close this issue, it's fixed for 6.0 and there are workarounds for 5.x.

**Comment by agolomoodysaada** [comment]
This still occurs in v6.7.0. Here's the stacktrace I'm getting

```
[2019-05-24T14:06:36,707][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [elk-elasticsearch-master-0] fatal error in thread [elasticsearch[elk-elasticsearch-master-0][masterService#updateTask][T#1]], exiting
java.lang.StackOverflowError: null
	at java.util.HashMap.hash(HashMap.java:339) ~[?:?]
	at java.util.LinkedHashMap.get(LinkedHashMap.java:440) ~[?:?]
	at java.util.Collections$UnmodifiableMap.get(Collections.java:1452) ~[?:?]
	at org.elasticsearch.index.mapper.Mapper$TypeParser$ParserContext.typeParser(Mapper.java:124) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.index.mapper.ObjectMapper$TypeParser.parseProperties(ObjectMapper.java:274) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.index.mapper.ObjectMapper$TypeParser.parseObjectOrDocumentTypeProperties(ObjectMapper.java:200) ~[elasticsearch-6.7.0.jar:6.7.0]
... (repeat the last few lines over and over)
```
